test/Npgsql.Tests/ConnectionTests.cs

```diff
@@ -1559,6 +1559,102 @@ public async Task TimeoutDuringAuthentication()
                     .With.InnerException.TypeOf<TimeoutException>());
         }
 
+        [Test]
+        public async Task Physical_open_callback_sync()
+        {
+            await using var defaultConn = await OpenConnectionAsync();
+            await using var _ = await CreateTempTable(defaultConn, "ID INTEGER", out var table);
+
+            using var __ = CreateTempPool(ConnectionString, out var connectionString);
+            using var conn = new NpgsqlConnection(connectionString);
+            conn.PhysicalOpenCallback = connector =>
+            {
+                using var cmd = connector.CreateCommand($"INSERT INTO \"{table}\" VALUES(1)");
+                cmd.ExecuteNonQuery();
+                return new ValueTask();
+            };
+
+            conn.Open();
+
+            var rowsCount = (long)(await conn.ExecuteScalarAsync($"SELECT COUNT(*) FROM \"{table}\""))!;
+            Assert.AreEqual(1, rowsCount);
+        }
+
+        [Test]
+        public async Task Physical_open_callback_async()
+        {
+            await using var defaultConn = await OpenConnectionAsync();
+            await using var _ = await CreateTempTable(defaultConn, "ID INTEGER", out var table);
+
+            using var __ = CreateTempPool(ConnectionString, out var connectionString);
+            await using var conn = new NpgsqlConnection(connectionString);
+            conn.PhysicalOpenCallback = async connector =>
+            {
+                using var cmd = connector.CreateCommand($"INSERT INTO \"{table}\" VALUES(1)");
+                await cmd.ExecuteNonQueryAsync();
+            };
+
+            await conn.OpenAsync();
+
+            var rowsCount = (long)(await conn.ExecuteScalarAsync($"SELECT COUNT(*) FROM \"{table}\""))!;
+            Assert.AreEqual(1, rowsCount);
+        }
+
+        [Test]
+        public async Task Physical_open_callback_async_over_sync()
+        {
+            using var _ = CreateTempPool(ConnectionString, out var connectionString);
+            await using var conn = new NpgsqlConnection(connectionString);
+            conn.PhysicalOpenCallback = async connector =>
+            {
+                await Task.Delay(50);
+            };
+
+            Assert.Throws<InvalidOperationException>(conn.Open);
+        }
+
+        [Test]
+        public async Task Physical_open_callback_throws()
+        {
+            PhysicalOpenCallback callback = _ => throw new NotImplementedException();
+
+            using var _ = CreateTempPool(ConnectionString, out var connectionString);
+            await using var conn = new NpgsqlConnection(connectionString);
+            conn.PhysicalOpenCallback = callback;
+
+            Assert.ThrowsAsync<NotImplementedException>(conn.OpenAsync);
+
+            if (IsMultiplexing)
+            {
+                await conn.CloseAsync();
```

Comment:
We shouldn't need to close the connection if OpenAsync throw an exception above... Is there some sort of bug here?

Reply:
Nope, there is no bug here - for some reason I thought we have to close the connection (or dispose it) whenever there is an exception while opening it.

Reply:
This is true (at least in theory) for exceptions out of operations once the connection is open, but if Open itself (or OpenAsync) throws the exception, we should be in exactly the same state as we were before Open was invoked.